cpp/src/join/sort_merge_join.cu

```diff
@@ -198,30 +202,71 @@ merge<LargerIterator, SmallerIterator>::operator()(rmm::cuda_stream_view stream,
+          std::unique_ptr<rmm::device_uvector<size_type>>>
+merge<LargerIterator, SmallerIterator>::operator()(rmm::cuda_stream_view stream,
+                                                   rmm::device_async_resource_ref mr)
+{
+  auto temp_mr              = cudf::get_current_device_resource_ref();
+  auto const larger_numrows = larger.num_rows();
+  auto smaller_dv_ptr       = cudf::table_device_view::create(smaller, stream);
+  auto larger_dv_ptr        = cudf::table_device_view::create(larger, stream);
+  auto list_lex_preprocess  = [stream](table_view const& table) {
+    std::vector<detail::dremel_data> dremel_data;
+    auto const num_list_columns = std::count_if(
+      table.begin(), table.end(), [](auto const& col) { return col.type().id() == type_id::LIST; });
+    auto dremel_device_views =
+      detail::make_empty_host_vector<detail::dremel_device_view>(num_list_columns, stream);
+    for (auto const& col : table) {
+      if (col.type().id() == type_id::LIST) {
+        dremel_data.push_back(detail::get_comparator_data(col, {}, false, stream));
+        dremel_device_views.push_back(dremel_data.back());
+      }
+    }
+    auto d_dremel_device_views = detail::make_device_uvector(
+      dremel_device_views, stream, cudf::get_current_device_resource_ref());
+    return std::pair(std::move(dremel_data), std::move(d_dremel_device_views));
+  };
+  auto [smaller_dremel, smaller_dremel_dv] = list_lex_preprocess(smaller);
+  auto [larger_dremel, larger_dremel_dv]   = list_lex_preprocess(larger);
+
+  rmm::device_scalar<bound_type> d_lb_type(bound_type::LOWER, stream, temp_mr);
+  row_comparator lb_comp(
+    *larger_dv_ptr, *smaller_dv_ptr, larger_dremel_dv, smaller_dremel_dv, d_lb_type.data());
+
+  // naive: iterate through larger table and binary search on smaller table
+  auto match_counts = matches_per_row(stream, temp_mr);
+
   auto count_matches_it = thrust::transform_iterator(
-    match_counts.begin(),
+    match_counts->begin(),
     cuda::proclaim_return_type<size_type>([] __device__(auto c) -> size_type { return c != 0; }));
   auto const count_matches =
     thrust::reduce(rmm::exec_policy(stream), count_matches_it, count_matches_it + larger_numrows);
   rmm::device_uvector<size_type> nonzero_matches(count_matches, stream, temp_mr);
   thrust::copy_if(rmm::exec_policy_nosync(stream),
                   thrust::counting_iterator(0),
                   thrust::counting_iterator(0) + larger_numrows,
-                  match_counts.begin(),
+                  match_counts->begin(),
                   nonzero_matches.begin(),
                   cuda::std::identity{});
 
   thrust::exclusive_scan(
-    rmm::exec_policy(stream), match_counts.begin(), match_counts.end(), match_counts.begin());
-  auto const total_matches = match_counts.back_element(stream);
+    rmm::exec_policy(stream), match_counts->begin(), match_counts->end(), match_counts->begin());
+  auto const total_matches = match_counts->back_element(stream);
```

Comment:
```suggestion
    rmm::exec_policy_nosync(stream), match_counts->begin(), match_counts->end(), match_counts->begin());
  auto const total_matches = match_counts->back_element(stream);
```
Could we use the async policy here? The subsequent `back_element` will likely synchronize the stream anyway, so an additional sync here might be redundant.